add / fix JSDoc annotations

mildsunrise opened the pull request:
finally, after #216 I think I have everything ready to add / fix the JSDoc comments on the code.
I'm pretty happy with the results, very very few errors remain ^^
and I found many typos along the way (#215)

the rationale is the same than for https://github.com/medooze/semantic-sdp-js/pull/28. our current typings are so outdated they're likely misleading people at this point rather than helping. so, delete them and autogenerate them from the JSDoc comments in our code.

**not ready for review yet.** there are still some things I'd like to take out of this PR first, to ease review (a bugfix and some needed refactors)

mildsunrise commented:
note to self in case I hit this problem again. the tests are failing because... we're using a very old version of TAP which uses the `esm` package to simulate ESM support back when Node.js did not have it yet.

this package will, by default, use a "heuristic" to figure out if a file is ESM or CJS. the heuristic is to search for the words `import`, `export`, and so on. the module DOES NOT care if they only occur inside a comment and not in actual code. amazing. so files are now being wrongly detected as ESM and transpiled accordingly, which causes the syntax issues.

this would be solved by upgrading TAP to v15 (TAP version in other repos is 16). HOWEVER if I do that, the code breaks because the `test.done()` function no longer exists. I tried to find that function in [documentation](https://node-tap.org/api/); it is not mentioned anywhere. [neither in older versions](https://github.com/tapjs/tapjs/blob/v11.1.5/docs/api/index.md). in the [list of breaking changes for v15](https://github.com/tapjs/tapjs/blob/v15.0.0/docs/src/content/changelog/index.md#150---2021-03-30) it is also not mentioned. nevertheless it seems this [is an alias to `test.end()`]() that they decided to remove when they made their refactor and extracted functionality into libtap.

so I'll open another PR to upgrade to TAP v16 to match the other repos, and then mass rename `test.done()` to `test.end()`.

murillo128 commented on lib/IncomingStream.js:
```diff
@@ -107,28 +124,17 @@ class IncomingStream extends Emitter
-		for (const [trackId,stat] of Object.entries(stats))
-			//Get stat
-			stats[trackId] = await stat;
+		const promises = [...this.tracks.values()].map(track =>
```
please make it a bit easier to understand, even if the code is longer. Or add some comments

mildsunrise replied:
i took this out to https://github.com/medooze/media-server-node/pull/225, marking this thread resolved.

murillo128 commented on lib/SDPManagerPlanB.js:
```diff
@@ -68,12 +76,13 @@ class SDPManagerPlanB extends SDPManager
 		const renegotiate = () => {
 			//Check if we already need to renegotiate
-			if (!this.renegotiationNeeded && ["initial","stable"].includes(this.state))
+			if (!this.renegotiationNeeded && (this.state === 'initial' || this.state === 'stable'))
```
why?

mildsunrise replied:
i took this out to #224, marking this thread resolved.

mildsunrise replied:
i took this out to https://github.com/medooze/media-server-node/pull/225, marking this thread resolved.

murillo128 commented on lib/IncomingStream.js:
```diff
@@ -353,7 +318,7 @@ class IncomingStream extends Emitter
 					{
 						//Create single incoming source
-						const source = new Native.RTPIncomingSourceGroupShared(type,this.transport.GetTimeService());
+						const source = SharedPointer(new Native.RTPIncomingSourceGroupShared(type,this.transport.GetTimeService()));
```
Why are you doing this?

The Native objects are wrapped into a shared pointer automatically:

https://github.com/medooze/media-server-node/blob/master/lib/SharedPointer.js#L61

mildsunrise replied:
i took this out to #224, marking this thread resolved.

mildsunrise commented on lib/ActiveSpeakerMultiplexer.js:
```diff
@@ -11,28 +30,24 @@ class ActiveSpeakerMultiplexer extends Emitter
 		
 		//Listen for speaker changes		
-		this.onactivespeakerchanged = (speakerId,multiplexeId) => {
-			/**
```
this PR is not yet ready for review, and the removing of event blocks was one of the things I wanted to take out of this PR. I'll explain the motivation in the separate PR.

mildsunrise replied:
i took this out to #223, marking this thread resolved.

mildsunrise commented:
rebased against latest master.
this is still not ready for review.

mildsunrise commented:
I've taken most stuff out of this PR, which greatly reduces the amount of stuff to review.
there should be no refactors here now, i've only left the annotations / comments.

murillo128 commented on lib/IncomingStreamTrackMirrored.js:
```diff
@@ -101,8 +117,6 @@ class IncomingStreamTrackMirrored extends Emitter
 	/**
 	 * Get stats for all encodings from the original track
-	 * 
-	 * @returns {Map<String,Object>} Map with stats by encodingId
```
why remove the return type info?

mildsunrise replied:
because it's incorrect :sweat_smile:

specifying the return type is optional, the type system can figure it out for you.
it's still good practice to specify it, but for simple functions which only call another function, it's not worth it imo.

murillo128 replied:
i am mostly worried about docs

mildsunrise replied:
typedoc shows precisely what the type system knows, regardless if you annotated it explicitly or not.
for example, in nal-extractor this function isn't annotated with a return type:

https://github.com/medooze/nal-extractor/blob/57fb85568b3a667d09b2dc26ede38497b872199c/lib/NalUnits.ts#L78-L83

but its documentation shows it precisely: https://medooze.github.io/nal-extractor/main/functions/NalUnits.parseNALU.html

are you okay with switching to typedoc to render the documentation?

murillo128 replied:
yes, I am ok

murillo128 reviewed (approved):
lgtm, minor comment

murillo128 commented:
last rebase and I will merge, I promise.. :)

mildsunrise commented:
rebased and added some last-minute annotations in https://github.com/medooze/media-server-node/pull/217/commits/85a28b7ae7fc4c394de354c1df83a0ec87de5fdb

murillo128 reviewed (approved):
lgtm
